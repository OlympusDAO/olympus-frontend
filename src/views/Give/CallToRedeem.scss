.call-to-redeem {
  display: flex;
  justify-content: center;
  background-color: #4fc4fa;
  color: #0e0333;
  border-radius: 8px;
  padding: 4px;
}
