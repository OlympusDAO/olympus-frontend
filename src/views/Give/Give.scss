#give-view {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  overflow: hidden;
  margin-top: 10px;

  .ohm-card {
    overflow-y: scrollable;
    .cell-align-end {
      display: flex;
      justify-content: flex-end;
    }

    .MuiTypography-body2 {
      margin-bottom: 0.5rem;
    }

    .MuiTypography-body1 {
      margin-bottom: 0.5rem;
    }

    .header {
      min-height: 3em;
    }

    .button-box {
      display: flex;
      justify-content: center;

      .learn-more-button {
        p {
          margin-bottom: 0px;
        }

        svg {
          margin-left: 4px;
        }
      }
    }
  }

  .causes-container {
    padding: 0px 10px 30px 10px;
  }

  .card-header {
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    vertical-align: middle;

    .give-education {
      display: flex;
      flex-direction: row;
      justify-content: space-between;
      align-items: center;
      align-content: center;
    }
  }
  .causes-body {
    display: flex;
    flex-direction: column;
    align-items: center;
    align-content: center;
    justify-content: center;
    width: 100%;

    .MuiGrid-root.MuiGrid-item {
      display: flex;
      justify-self: center !important;
    }

    .data-grid {
      justify-content: center;
      align-items: center;
      height: 100%;
    }

    .cause-card {
      height: auto;
      &.very-small {
        min-width: 200px;
      }
      min-width: 333px;
      // max-width: 773px;
      width: auto;
      // box-shadow: 0px 2px 28px rgba(0, 0, 0, 0.03);
      // background-color: #ffffff0e;
      // backdrop-filter: blur(60px);
      // -webkit-backdrop-filter: blur(60px);
      border-radius: 10px !important;
      display: flex;
      flex-direction: row;
      justify-content: flex-start;
      align-items: center;
      margin-bottom: 2rem;

      .cause-title {
        padding-left: 0rem;
        padding-bottom: 0rem;
      }

      .cause-title:hover {
        h5 {
          text-decoration: underline;
        }
      }

      .cause-image {
        width: 35%;
        height: 100%;
        justify-content: flex-start;
        padding: 30px;

        img {
          border-radius: 5%;
        }
      }

      .cause-content {
        height: 100%;
        width: 65%;
        display: flex;
        flex-direction: column;
        justify-content: space-between;
        padding-right: 30px;
        padding-top: 30px;
        padding-bottom: 30px;

        .MuiTypography-body2 {
          margin-top: 0.5rem;
        }

        .cause-header {
          justify-content: space-between;
          align-items: center;
        }

        .cause-body {
          padding-bottom: 16px;

          p {
            margin-bottom: 0px;
          }
        }
      }

      .view-details {
        padding-bottom: 0rem;

        .cause-link {
          display: flex;
          flex-direction: row;
          align-content: center;
          align-items: center;
          justify-content: space-between;
          color: #999999;
          font-size: small;
          border-radius: 12px;
          font-weight: 500;

          svg path {
            fill: #999999;
          }

          > p {
            margin-bottom: 0px;
          }
        }
      }

      .cause-link:hover {
        color: #f4d092;

        svg path {
          fill: #f4d092;
        }
      }

      .cause-misc-info {
        display: flex;
        flex-direction: row;
        justify-content: space-between;
        align-items: center;

        .cause-info-main-text {
          font-size: 1rem;

          .MuiTypography-body1 {
            margin-bottom: 0rem;
          }
        }

        .cause-info-bottom-text {
          font-size: 0.875rem;
          font-weight: 500;
        }
      }

      .cause-misc-info .MuiGrid-item {
        height: 40px;
        padding-left: 0px;
        align-items: center;
        justify-content: flex-start;
      }

      .cause-info-icon {
        margin-right: 0.2rem;
        align-self: center;
      }

      .cause-give-button {
        height: 40px;
        width: 12rem;
        margin-left: auto;
        margin-right: 0;
      }
    }
  }

  .donation-table {
    display: flex;
    flex-direction: row;

    .donation-row {
      display: flex;
      justify-content: flex-end;
      width: 100%;

      .donation-buttons {
        display: flex;
        justify-content: flex-end;

        .donation-lp-button {
          width: 25%;
          margin-right: 1rem;
        }
      }

      .recipient-divider {
        width: 100%;
        margin-top: 1rem;
        margin-bottom: 1rem;
      }
    }

    .MuiTypography-h6 {
      min-width: 33%;
      margin-right: 2rem;
    }

    .MuiTypography-body1 {
      min-width: 33%;
      margin-top: 0.5rem;
      margin-right: 2rem;
      margin-bottom: 0rem;
    }
  }

  .donation-table.smaller {
    .donation-buttons {
      justify-content: flex-start;
      margin-top: 1rem;

      .donation-lp-button {
        width: 40%;
      }
    }
  }
}

#give-view.medium {
  .causes-body {
    .cause-content {
      width: 50%;

      .cause-body {
        padding-bottom: 0px;
      }
    }

    .cause-misc-info {
      .cause-give-button {
        width: 100%;
        margin-left: 0rem;
      }
    }

    .cause-image {
      width: 50%;
    }
  }
}

#give-view.smaller {
  .causes-container {
    padding: 5px 0px 30px 0px;
    .card-header {
      padding: 0px 0px 0px 30px;
    }
  }
  .causes-body {
    .cause-card {
      flex-direction: column;
    }

    .cause-content {
      padding-top: 0px;
      padding-left: 30px;
      width: 100%;
    }

    .cause-image {
      width: 100%;
    }
  }
}

.give-subnav {
  display: flex;
  flex-direction: row;
  align-content: flex-start;

  max-width: 833px;
  padding-top: 0.5rem;
  padding-bottom: 0.5rem;
  padding-left: 1rem;

  .MuiLink-root.active {
    text-decoration: none;
  }

  .give-option {
    margin-right: 1rem;
  }

  .give-option.give-active {
    text-decoration: underline;
  }

  #give-sub-dash {
    display: flex;
    flex-direction: row;
  }

  #give-sub-dash:hover {
    color: #f4d092;
  }
}

.give-subnav.smaller {
  margin-left: 6px;

  #give-sub-dash {
    margin-left: -7px;
  }
}

.custom-recipient {
  display: flex;
  flex-direction: row;
  justify-content: flex-end;
  &.smaller {
    padding: 0px 20px 5px 0px;
  }
  padding: 0px 30px 5px 0px;

  .custom-give-button {
    height: 40px;
    width: 12rem;
  }
}

.give-education-graphics {
  display: flex;
  flex-direction: row;
  justify-content: center;
  align-content: center;
  align-items: center;

  .yield-graphic svg path {
    fill: none;
  }
}

.give-education-graphics.smaller {
  flex-direction: column;
}

.give-staked-balance {
  display: flex;
  flex-direction: row;
  justify-content: space-between;

  .MuiTypography-body2 {
    margin-top: 0.5rem;
  }
}

.yield-action-area {
  justify-content: space-around;
  align-items: center;
  height: auto;
  padding-bottom: 1rem;
}

.give-yield-title {
  display: flex;
  flex-direction: row;
}

.give-yield-modals {
  display: flex;
  flex-direction: row;

  .add-recipient-button {
    margin-right: 0.33em;
  }
}

.ohm-modal {
  position: absolute;
  left: 50%;
  top: 50%;
  // Use decimal values to fix anti-aliasing in Chrome. Ridiculous.
  transform: translate(-50.048%, -50.048%);

  .yield-header {
    align-items: center;
    display: flex;
    height: 44px;
    justify-content: center;
    margin-bottom: 30px;

    a {
      position: absolute;
      left: 1.7rem;
    }
  }

  .give-modal-header {
    display: flex;
    flex-direction: row;
  }

  .give-modal-alloc-tip {
    display: flex;
    flex-direction: row;

    .MuiBox-root {
      margin-bottom: 0.5rem;
    }
  }

  .modal-input {
    width: 100%;
    max-width: 100%;
    padding-bottom: 1rem;

    .MuiFormHelperText-root {
      color: red;
    }
  }

  .ohm-modal-submit {
    display: flex;
    align-items: center;
    width: 60%;
    height: 43px !important;
    margin-top: 1rem;
    margin-left: auto;
    margin-right: auto;

    button {
      height: 40px;
      width: 100%;
    }
  }
}

.ohm-modal.smaller {
  top: 0%;
  transform: translate(-50.048%, 0%);
}

.give-confirmation-details {
  .confirmation-sect-header {
    margin-bottom: 16px;
  }

  .details-row {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    align-items: center;

    .sohm-allocation-col {
      width: 30%;
    }

    .recipient-address-col {
      width: 30%;
      display: flex;
      flex-direction: column;
      align-items: flex-end;
      align-content: flex-end;
    }
  }
}

.give-confirmation-divider {
  margin-top: 16px;
  margin-bottom: 16px;

  .MuiDivider-root {
    background-color: #f4d092;
  }
}

.give-info {
  display: flex;
  flex-direction: row;
  justify-content: space-between;

  .subtext {
    font-weight: 400;
  }

  .give-info-deposit-box .text {
    margin-left: 30px;
  }

  .give-info-vault-box .text {
    margin-left: 30px;
  }

  .give-info-yield-box {
    .text {
      margin-left: 30px;
    }

    .receives-yield-icon path {
      fill: none;
    }
  }
}

.redeem-table {
  .MuiTableCell-body {
    font-size: 0.875rem;
    font-family: Square;
    font-weight: 500;
    line-height: 1;
    padding-left: 0rem;
    padding-bottom: 1rem;
  }
}

.arrow-graphic {
  svg path {
    fill: #999999;
  }
}

.yield-graphic {
  svg path {
    fill: none;
  }
}

.project {
  margin-top: 10px;
  // TODO look at whether we can make this less flakey by using rem instead of px
  .project-sidebar {
    padding-left: 30px;
    padding-right: 30px;
    padding-top: 30px;
    padding-bottom: 1.5rem;
    margin-bottom: 1rem;

    .cause-image {
      padding-bottom: 30px;
    }

    .project-goal .project-donated-icon {
      flex-direction: row;
      flex-wrap: nowrap;

      .subtext {
        color: #999999;
      }

      h6 {
        line-height: 1rem;
      }

      .MuiGrid-item {
        max-width: 20px;
        margin-right: 1rem;
      }
    }

    .project-countdown .countdown-container {
      display: flex;
      flex-direction: row;
      justify-content: center;

      .countdown-object {
        display: flex;
        flex-direction: row;
        align-items: center;
        justify-content: center;

        svg {
          padding-right: 0.3rem;
        }
      }

      .project-countdown-text {
        display: flex;
        align-content: center;
        justify-content: center;
        padding-top: 0.3rem;

        .cause-info-bottom-text {
          color: #999999;
        }
      }
    }

    .donors-title {
      padding-bottom: 30px;
    }
  }

  .project-intro {
    padding-bottom: 30px;
  }

  .project-link {
    padding-top: 0.4rem;
  }

  .project-info {
    padding-left: 30px;
    padding-right: 30px;
    padding-top: 30px;
    padding-bottom: 1rem;
    margin-left: 2rem;
    height: auto;
    width: 100%;

    .project-about-header {
      padding-bottom: 30px;
    }

    p {
      margin-top: 0rem;
      font-size: 0.875rem;
      line-height: 20px;
    }
  }

  .project-donated {
    .project-donated-icon {
      display: inline-flex;
      align-items: center;

      h6 {
        line-height: 1rem;
      }
    }

    .subtext {
      color: #999999;
    }
  }

  .project-completion {
    text-align: right;

    .project-completion-icon {
      display: inline-flex;
      align-items: center;

      h6 {
        max-width: 3.3rem;
        line-height: 1rem;
      }
    }

    .subtext {
      color: #999999;
    }
  }

  .project-goal-progress {
    padding-top: 0.5rem;
    padding-bottom: 1.5rem;
  }

  .project-give-button {
    padding-top: 1.5rem;

    button {
      width: 100%;
    }
  }

  .MuiPaper-root {
    border-radius: 10px;
  }

  .MuiLinearProgress-determinate {
    height: 1rem;
    border-radius: 3px;
  }

  .MuiLinearProgress-barColorPrimary {
    background: rgb(112, 139, 150);
    background: linear-gradient(269deg, rgba(112, 139, 150, 1) 0%, rgba(247, 251, 231, 1) 100%);
  }
}

.project-container {
  margin-bottom: 7.5rem;

  .very-small {
    .project-info {
      margin-left: 0;
    }

    .visual-info-bottom {
      flex-direction: column;

      .project-give-button {
        padding-top: 1.5rem;
      }
    }
  }

  .smaller {
    .project-info {
      margin-left: 0;
    }

    .project-visual-info {
      flex-direction: column;

      .cause-image {
        width: 100%;
        margin-right: 0px;
      }

      .goal-graphics {
        width: 100%;
      }

      .visual-info-bottom {
        display: flex;
        flex-direction: row;
        align-items: center;

        .project-give-button {
          padding-top: 0px;
        }

        .project-countdown {
          width: 50%;
        }
      }
    }
  }

  .medium {
    .project-info {
      margin-left: 0;
    }

    .project-visual-info {
      display: flex;
      flex-direction: row;
      justify-content: space-between;
      align-items: center;

      .cause-image {
        width: 50%;
        margin-right: 30px;
      }

      .goal-graphics {
        width: 50%;
      }
    }
  }
}

.yield-recipients-empty {
  align-items: center;
}

@supports (-webkit-backdrop-filter: none) or (backdrop-filter: none) {
  .modal-container {
    background: rgba(100, 100, 100, 0.1) !important;
    backdrop-filter: blur(33px) !important;
    -webkit-backdrop-filter: blur(33px) !important;
    overflow-y: scroll;
    .ohm-card {
      opacity: 1 !important;
      height: auto;
    }
    .ohm-modal {
      max-width: 688px;
      opacity: 0.9;
    }
  }
}

/* slightly transparent fallback for Firefox (not supporting backdrop-filter) */
@supports not ((-webkit-backdrop-filter: none) or (backdrop-filter: none)) {
  .modal-container {
    background: rgba(100, 100, 100, 0.95);
    overflow-y: scroll;
    .ohm-card {
      height: auto;
    }
  }

  .ohm-modal {
    max-width: 688px;
    opacity: 0.9;
  }
}

.subnav-paper {
  max-width: 833px;
  margin-bottom: 10rem;
  &.mobile {
    width: 100%;
  }
}

.project-content {
  li {
    // Same as standard
    line-height: 20px;
  }
}
