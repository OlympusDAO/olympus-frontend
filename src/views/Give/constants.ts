export const NEW_DEPOSIT = "-1";
/**
 * We set the decimals amount to 9 to try to limit any precision issues with
 * sOHM and gOHM conversions on the contract side
 */
export const GIVE_MAX_DECIMALS = 9;
/**
 * An options preset to be used with `DecimalBigNumber` that enforces a reasonable number
 * decimal places to limit precision issues with sOHM and gOHM conversions.
 */
export const GIVE_MAX_DECIMAL_FORMAT = { decimals: GIVE_MAX_DECIMALS, format: false };
