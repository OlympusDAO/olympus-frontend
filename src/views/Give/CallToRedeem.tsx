import "./CallToRedeem.scss";

import { Trans } from "@lingui/macro";
import { Box, Typography } from "@mui/material";

export const CallToRedeem = () => {
  return (
    <Box className="call-to-redeem">
      <Typography variant="h6">
        <Trans>You have redeemable yield on the old contract</Trans>
      </Typography>
    </Box>
  );
};
