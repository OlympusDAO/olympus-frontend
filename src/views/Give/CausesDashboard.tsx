import "./Give.scss";

import { t, Trans } from "@lingui/macro";
import { Box, Button, Container, Typography, Zoom } from "@material-ui/core";
import useMediaQuery from "@material-ui/core/useMediaQuery";
import { BigNumber } from "bignumber.js";
import { useMemo, useState } from "react";
import { useSelector } from "react-redux";
import { useLocation } from "react-router-dom";
import { useUIDSeed } from "react-uid";
import { GiveHeader } from "src/components/GiveProject/GiveHeader";
import ProjectCard, { ProjectDetailsMode } from "src/components/GiveProject/ProjectCard";
import { NetworkId } from "src/constants";
import { EnvHelper } from "src/helpers/Environment";
import { useAppDispatch } from "src/hooks";
import { useWeb3Context } from "src/hooks/web3Context";
import { IAccountSlice } from "src/slices/AccountSlice";
import { IAppData } from "src/slices/AppSlice";
import { ACTION_GIVE, changeGive, changeMockGive, isSupportedChain } from "src/slices/GiveThunk";
import { IPendingTxn } from "src/slices/PendingTxnsSlice";
import { CancelCallback, SubmitCallback } from "src/views/Give/Interfaces";
import { RecipientModal } from "src/views/Give/RecipientModal";

import { error } from "../../slices/MessagesSlice";
import { GiveInfo } from "./GiveInfo";
import data from "./projects.json";

type State = {
  account: IAccountSlice;
  pendingTransactions: IPendingTxn[];
  app: IAppData;
};
export default function CausesDashboard() {
  const location = useLocation();
  const { provider, address, networkId } = useWeb3Context();
  const [isCustomGiveModalOpen, setIsCustomGiveModalOpen] = useState(false);
  const isSmallScreen = useMediaQuery("(max-width: 600px)");
  const isMediumScreen = useMediaQuery("(max-width: 980px)") && !isSmallScreen;
  const { projects } = data;

  // We use useAppDispatch here so the result of the AsyncThunkAction is typed correctly
  // See: https://stackoverflow.com/a/66753532
  const dispatch = useAppDispatch();
  const seed = useUIDSeed();

  const donationInfo = useSelector((state: State) => {
    return networkId === NetworkId.TESTNET_RINKEBY && EnvHelper.isMockSohmEnabled(location.search)
      ? state.account.mockGiving && state.account.mockGiving.donationInfo
      : state.account.giving && state.account.giving.donationInfo;
  });

  const totalDebt = useSelector((state: State) => {
    return networkId === NetworkId.TESTNET_RINKEBY && EnvHelper.isMockSohmEnabled(location.search)
      ? state.account.mockRedeeming && state.account.mockRedeeming.recipientInfo.totalDebt
      : state.account.redeeming && state.account.redeeming.recipientInfo.totalDebt;
  });

  const renderProjects = useMemo(() => {
    return projects.map(project => {
      return <ProjectCard key={seed(project.title)} project={project} mode={ProjectDetailsMode.Card} />;
    });
  }, [projects]);

  const handleCustomGiveButtonClick = () => {
    setIsCustomGiveModalOpen(true);
  };

  const handleCustomGiveModalSubmit: SubmitCallback = async (
    walletAddress: string,
    depositAmount: BigNumber,
    depositAmountDiff?: BigNumber,
  ) => {
    if (depositAmount.isEqualTo(new BigNumber(0))) {
      return dispatch(error(t`Please enter a value!`));
    }

    // If reducing the amount of deposit, withdraw
    if (networkId === NetworkId.TESTNET_RINKEBY && EnvHelper.isMockSohmEnabled(location.search)) {
      await dispatch(
        changeMockGive({
          action: ACTION_GIVE,
          value: depositAmount.toFixed(),
          recipient: walletAddress,
          provider,
          address,
          networkID: networkId,
          version2: false,
          rebase: false,
        }),
      );
    } else {
      await dispatch(
        changeGive({
          action: ACTION_GIVE,
          value: depositAmount.toFixed(),
          recipient: walletAddress,
          provider,
          address,
          networkID: networkId,
          version2: false,
          rebase: false,
        }),
      );
    }

    setIsCustomGiveModalOpen(false);
  };

  const handleCustomGiveModalCancel: CancelCallback = () => {
    setIsCustomGiveModalOpen(false);
  };

  return (
    <Container
      style={{
        paddingLeft: isSmallScreen ? "0" : "3.3rem",
        paddingRight: isSmallScreen ? "0" : "3.3rem",
        display: "flex",
        justifyContent: "center",
      }}
    >
      <Box className={isSmallScreen ? "subnav-paper mobile" : "subnav-paper"}>
        <GiveHeader
          isSmallScreen={isSmallScreen}
          isVerySmallScreen={false}
          totalDebt={new BigNumber(totalDebt)}
          networkId={networkId}
        />
        <div
          id="give-view"
          className={`${isMediumScreen && "medium"}
          ${isSmallScreen && "smaller"}`}
        >
          <Zoom in={true}>
            <Box className={`ohm-card secondary causes-container`}>
              {!isSupportedChain(networkId) ? (
                <Typography variant="h6">
                  Note: You are currently using an unsupported network. Please switch to Ethereum to experience the full
                  functionality.
                </Typography>
              ) : (
                <></>
              )}
              <div className="causes-body">
                <Box className="data-grid">{renderProjects}</Box>
              </div>
              <div className={isSmallScreen ? "custom-recipient smaller" : "custom-recipient"}>
                <Button
                  variant="contained"
                  color="primary"
                  className="custom-give-button"
                  onClick={() => handleCustomGiveButtonClick()}
                  disabled={!address}
                >
                  <Typography variant="h6" style={{ marginBottom: "0px" }}>
                    <Trans>Custom Recipient</Trans>
                  </Typography>
                </Button>
              </div>
              <RecipientModal
                isModalOpen={isCustomGiveModalOpen}
                callbackFunc={handleCustomGiveModalSubmit}
                cancelFunc={handleCustomGiveModalCancel}
              />
            </Box>
          </Zoom>
          <Zoom in={true}>
            <GiveInfo />
          </Zoom>
        </div>
      </Box>
    </Container>
  );
}
