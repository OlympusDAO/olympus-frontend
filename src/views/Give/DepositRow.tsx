import "./YieldRecipients.scss";

import { t } from "@lingui/macro";
import { Grid, Tooltip, Typography } from "@material-ui/core";
import { useTheme } from "@material-ui/core/styles";
import useMediaQuery from "@material-ui/core/useMediaQuery";
import { SecondaryButton } from "@olympusdao/component-library";
import { BigNumber } from "bignumber.js";
import { useState } from "react";
import { useDispatch } from "react-redux";
import { useLocation } from "react-router-dom";
import { NetworkId } from "src/constants";
import { Environment } from "src/helpers/environment/Environment/Environment";
import { useWeb3Context } from "src/hooks/web3Context";
import { SubmitCallback } from "src/views/Give/Interfaces";

import { Project } from "../../components/GiveProject/project.type";
import { ACTION_GIVE_EDIT, ACTION_GIVE_WITHDRAW, changeGive, changeMockGive } from "../../slices/GiveThunk";
import { error } from "../../slices/MessagesSlice";
import { ManageDonationModal, WithdrawSubmitCallback } from "./ManageDonationModal";
import data from "./projects.json";

interface IUserDonationInfo {
  date: string;
  deposit: string;
  recipient: string;
  yieldDonated: string;
}

interface DepositRowProps {
  depositObject: IUserDonationInfo;
}

export const DepositTableRow = ({ depositObject }: DepositRowProps) => {
  const location = useLocation();
  const dispatch = useDispatch();
  const { provider, address, networkId } = useWeb3Context();
  const { projects } = data;
  const projectMap = new Map(projects.map(i => [i.wallet, i] as [string, Project]));
  const [isManageModalOpen, setIsManageModalOpen] = useState(false);
  const theme = useTheme();
  const isSmallScreen = useMediaQuery(theme.breakpoints.down("xs"));
  const isMediumScreen = useMediaQuery(theme.breakpoints.down("md"));

  const getRecipientTitle = (address: string): string => {
    const project = projectMap.get(address);
    if (!project) return isMediumScreen || isSmallScreen ? "Custom" : "Custom Recipient";

    if (!project.owner)
      return isSmallScreen && project.title.length > 16 ? project.title.substring(0, 16) + "..." : project.title;

    return project.owner + " - " + project.title;
  };

  const handleManageModalCancel = () => {
    setIsManageModalOpen(false);
  };

  const handleEditModalSubmit: SubmitCallback = async (
    walletAddress,
    eventSource,
    depositAmount,
    depositAmountDiff,
  ) => {
    if (!depositAmountDiff) {
      return dispatch(error(t`Please enter a value!`));
    }

    if (depositAmountDiff.isEqualTo(new BigNumber(0))) return;

    // If on Rinkeby and using Mock Sohm, use changeMockGive async thunk
    // Else use standard call
    if (networkId === NetworkId.TESTNET_RINKEBY && Environment.isMockSohmEnabled(location.search)) {
      await dispatch(
        changeMockGive({
          action: ACTION_GIVE_EDIT,
          value: depositAmountDiff.toFixed(),
          recipient: walletAddress,
          provider,
          address,
          networkID: networkId,
          version2: false,
          rebase: false,
          eventSource,
        }),
      );
    } else {
      await dispatch(
        changeGive({
          action: ACTION_GIVE_EDIT,
          value: depositAmountDiff.toFixed(),
          recipient: walletAddress,
          provider,
          address,
          networkID: networkId,
          version2: false,
          rebase: false,
          eventSource,
        }),
      );
    }

    setIsManageModalOpen(false);
  };

  // If on Rinkeby and using Mock Sohm, use changeMockGive async thunk
  // Else use standard call
  const handleWithdrawModalSubmit: WithdrawSubmitCallback = async (walletAddress, eventSource, depositAmount) => {
    // Issue withdrawal from smart contract
    if (networkId === NetworkId.TESTNET_RINKEBY && Environment.isMockSohmEnabled(location.search)) {
      await dispatch(
        changeMockGive({
          action: ACTION_GIVE_WITHDRAW,
          value: depositAmount.toFixed(),
          recipient: walletAddress,
          provider,
          address,
          networkID: networkId,
          version2: false,
          rebase: false,
          eventSource,
        }),
      );
    } else {
      await dispatch(
        changeGive({
          action: ACTION_GIVE_WITHDRAW,
          value: depositAmount.toFixed(),
          recipient: walletAddress,
          provider,
          address,
          networkID: networkId,
          version2: false,
          rebase: false,
          eventSource,
        }),
      );
    }

    setIsManageModalOpen(false);
  };

  return (
    <Grid container alignItems="center" spacing={2}>
      {!isSmallScreen && (
        <Grid item xs={2}>
          <Typography variant="body1">{depositObject.date}</Typography>
        </Grid>
      )}
      <Grid item xs={4} sm={3}>
        <Tooltip title={depositObject.recipient} arrow>
          <Typography variant="body1">{getRecipientTitle(depositObject.recipient)}</Typography>
        </Tooltip>
      </Grid>
      {!isSmallScreen && (
        <Grid item xs={2} style={{ textAlign: "right" }}>
          <Typography variant="body1">{parseFloat(depositObject.deposit).toFixed(2)} sOHM</Typography>
        </Grid>
      )}
      <Grid item xs={4} sm={2} style={{ textAlign: "right" }}>
        <Typography variant="body1">{parseFloat(depositObject.yieldDonated).toFixed(2)} sOHM</Typography>
      </Grid>
      <Grid item xs={4} sm={3} style={{ textAlign: "right" }}>
        <SecondaryButton onClick={() => setIsManageModalOpen(true)} size="small" fullWidth>
          Manage
        </SecondaryButton>
      </Grid>
      <ManageDonationModal
        isModalOpen={isManageModalOpen}
        submitEdit={handleEditModalSubmit}
        submitWithdraw={handleWithdrawModalSubmit}
        cancelFunc={handleManageModalCancel}
        currentWalletAddress={depositObject.recipient}
        currentDepositAmount={new BigNumber(depositObject.deposit)}
        depositDate={depositObject.date}
        yieldSent={depositObject.yieldDonated}
        project={projectMap.get(depositObject.recipient)}
        key={"manage-modal-" + depositObject.recipient}
        eventSource={"My Donations"}
      />
    </Grid>
  );
};
