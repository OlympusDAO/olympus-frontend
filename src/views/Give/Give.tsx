import "./Give.scss";

import { t } from "@lingui/macro";
import { Grid, Link, Typography } from "@mui/material";
import { useTheme } from "@mui/material/styles";
import useMediaQuery from "@mui/material/useMediaQuery";
import { Paper, Tab, TabPanel, Tabs } from "@olympusdao/component-library";
import { useState } from "react";
import { Outlet, Route, Routes } from "react-router";
import { NavLink as RouterLink } from "react-router-dom";
import { isSupportedChain } from "src/helpers/GiveHelpers";
import { useV1RedeemableBalance } from "src/hooks/useGiveInfo";
import { useWeb3Context } from "src/hooks/web3Context";
import { ChangeAssetType } from "src/slices/interfaces";

import { CallToRedeem } from "./CallToRedeem";
import CausesDashboard from "./CausesDashboard";
import { GiveInfo } from "./GiveInfo";
import { GohmToggle } from "./GohmToggle";
import GrantInfo from "./GrantInfo";
import GrantsDashboard from "./GrantsDashboard";
import ProjectInfo from "./ProjectInfo";
import RedeemYield from "./RedeemYield";
import YieldRecipients from "./YieldRecipients";

function Give({ selectedIndex = 0 }) {
  const [giveAssetType, setGiveAssetType] = useState<"sOHM" | "gOHM">("sOHM");

  const { address, networkId } = useWeb3Context();

  const v1RedeemableBalance = useV1RedeemableBalance(address);
  const hasV1Assets = v1RedeemableBalance.data && v1RedeemableBalance.data != "0.0";

  const theme = useTheme();
  const isBreakpointXS = useMediaQuery(theme.breakpoints.down("sm"));

  const changeGiveAssetType: ChangeAssetType = (checked: boolean) => {
    setGiveAssetType(checked ? "gOHM" : "sOHM");
  };

  return (
    <>
      <Routes>
        <Route
          path="/"
          element={
            <Grid container direction="column" alignItems="center" className="give-container">
              <Grid item xs={12} sm={10} md={10} lg={8}>
                <Paper
                  headerText={t`Give`}
                  childPaperBackground={true}
                  fullWidth
                  className="no-container-padding"
                  zoom={false}
                >
                  {!isSupportedChain(networkId) ? (
                    <Typography variant="h6">
                      Note: You are currently using an unsupported network. Please switch to Ethereum to experience the
                      full functionality.
                    </Typography>
                  ) : (
                    <></>
                  )}
                  {hasV1Assets && <CallToRedeem />}
                  <Tabs
                    centered
                    value={selectedIndex}
                    className={`give-tab-buttons ${isBreakpointXS ? `give-tab-buttons-xs` : ``}`}
                    aria-label="stake tabs"
                    TabIndicatorProps={{ style: { display: "none" } }}
                  >
                    <Link to="/give" component={RouterLink} end>
                      <Tab label={t`Causes`} />
                    </Link>
                    <Link to="/give/grants" component={RouterLink}>
                      <Tab label={t`Grants`} />
                    </Link>
                    <Link to="/give/donations" component={RouterLink}>
                      <Tab label={t`My Donations`} style={{ whiteSpace: "nowrap" }} />
                    </Link>
                    <Link to="/give/redeem" component={RouterLink}>
                      <Tab label={t`Redeem`} />
                    </Link>
                  </Tabs>
                  <Outlet />
                </Paper>
                <GiveInfo />
              </Grid>
            </Grid>
          }
        >
          <Route
            index
            element={
              <TabPanel value={0} index={0}>
                <GohmToggle giveAssetType={giveAssetType} changeAssetType={changeGiveAssetType} />
                <CausesDashboard giveAssetType={giveAssetType} changeAssetType={changeGiveAssetType} />
              </TabPanel>
            }
          />

          <Route
            path="grants"
            element={
              <TabPanel value={1} index={1}>
                <GohmToggle giveAssetType={giveAssetType} changeAssetType={changeGiveAssetType} />
                <GrantsDashboard giveAssetType={giveAssetType} changeAssetType={changeGiveAssetType} />
              </TabPanel>
            }
          />

          <Route
            path="donations"
            element={
              <TabPanel value={2} index={2}>
                {/* We have a button to switch tabs in this child component, so need to pass the handler. */}
                <YieldRecipients giveAssetType={giveAssetType} changeAssetType={changeGiveAssetType} />
              </TabPanel>
            }
          />

          <Route
            path="redeem"
            element={
              <TabPanel value={3} index={3}>
                <RedeemYield />
              </TabPanel>
            }
          />
        </Route>

        <Route
          path="/projects/:slug"
          element={<ProjectInfo giveAssetType={giveAssetType} changeAssetType={changeGiveAssetType} />}
        />

        <Route
          path="/grants/:slug"
          element={<GrantInfo giveAssetType={giveAssetType} changeAssetType={changeGiveAssetType} />}
        />
      </Routes>
    </>
  );
}

export default Give;
