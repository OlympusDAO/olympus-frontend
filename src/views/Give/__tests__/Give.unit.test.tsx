import { render } from "../../../testUtils";
import Give from "../Give";

describe("<Give/>", () => {
  it("should render component", () => {
    const { container } = render(<Give />);
    expect(container).toMatchSnapshot();
  });
});
