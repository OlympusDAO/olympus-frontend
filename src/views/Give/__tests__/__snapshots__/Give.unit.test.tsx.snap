// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Give/> should render component 1`] = `
<div>
  <div
    class="give-view"
  >
    <div
      class="MuiPaper-root ohm-card secondary false MuiPaper-elevation0 MuiPaper-rounded"
      style="transform: none; webkit-transition: transform 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: transform 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    >
      <div
        class="stake-wallet-notification"
      >
        <div
          class="wallet-menu"
          id="wallet-menu"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-root-1 makeStyles-root-2 undefined MuiButton-containedPrimary MuiButton-containedSizeLarge MuiButton-sizeLarge MuiButton-disableElevation"
            style="font-size: 1.2857rem;"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              Connect Wallet
            </span>
          </button>
        </div>
        <h6
          class="MuiTypography-root MuiTypography-h6"
        >
          Connect your wallet to give or redeem OHM
        </h6>
      </div>
    </div>
  </div>
</div>
`;
